---
title: "Drp1 isoform analysis with UMI-deduplication"
format:
  html:
    embed-resources: true
    toc: true
    code-fold: true
    code-overflow: wrap
execute: 
  cache: true
  warning: false
  error: false
---

## Preprossing pipeline
### Base calling and demultiplexing

Use output from fastq_pass files (no trim, SUP mode).


### Run Flexiplex to remove barcode and UMI
* 1, First step to get UMI and attach to the beginning of read. 
* 2, Second step to remove barcode.
* 3, Then need to move UMI to the end of read.
* Flexiplex is able to get 85% reads after UMI and BC extraction. 

```
zcat ${raw_reads} | \
flexiplex -x $TSO_REV_CMP -u $UMI_REV_CMP -x $UMI_LEFT_FLANK_REV_CMP \
  -b "" -k "?" \
  -f 2 -e 1 -p 8 -n "UMI_extraction" | \
flexiplex -x $BC_LEFT_FLANK_FWD -b $BC_SEQ -x $BC_RIGHT_FLANK_FWD -k "?" -f 1 -p 8 -n "BC_extraction"  | \
sed 's/\(\?_#?\)\(_[^#]*\)\(#.*\)/\1\3\2/' > ${base}_fp.fastq
```

### Run UMI-tools to deduplicate UMI

* Map to human (GRCh38) genome using the code below.
* Then UMI-tools to remove duplicates based on UMI and coordinate (only look at the chromosome with DNM1L).

```
paftools.js gff2bed $gtf > ${name}_anno.bed
minimap2 -ax splice -I 1000G -t $num_cores --junc-bed ${name}_anno.bed $genome $raw_reads | samtools sort -@ $num_cores -O BAM -o ${name}_sorted.bam
samtools index -@ $num_cores ${name}_sorted.bam
umi_tools dedup -I ${name}_sorted.bam --chrom=${chr} --output-stats=${name} --edit-distance-threshold=2 --method=directional -S ${name}_dedup.bam

```

### Run Bambu

Run Bambu with gencode annotation (NCBI Refseq GCA_000001405.15 for human) with NDR=0.2 (relative strigent detection of novel isoform).

```         
Rscript --vanilla bambu_analysis.R Drp1_human_dge/
```

## Visualization of prelimary results 

```{r, warning=FALSE, message=FALSE}

library(ggfortify)
library(tidyverse)
library(bambu)
library(pheatmap)
library(DT)
library(ggh4x)
library(cowplot)
library(ComplexHeatmap)
library(RColorBrewer)

```

### Human data, initial QC stats
```{r}
se2 <- readRDS("Drp1_human_dge/default/se1.rds")[,-11] # remove RNA standard

group2 <- factor(c('CERA','CERA','CERA','LV','LV','LV','LV','CL2','CL2','CL2'))

```

```{r, fig.height=5, fig.width=12, dev=c('png', 'pdf')}
#| fig-cap: "Fig1 Sequencing stats of each sample"
library.size2 <- colSums(assays(se2)$counts)
gene.count2 <- colSums(assays(se2)$count[rowData(se2)$GENEID == 'DNM1L',])
capture.rate2 <- gene.count2/library.size2

# hist(capture.rate2, main = 'capture rate in chr12')

anno.col2 <- data.frame(capture.rate = capture.rate2,
                        count_DNM1L = gene.count2, 
                       count_chr12 = library.size2,
                       group = group2)
write.csv(anno.col2, 'human_stats.csv')

anno.col2 %>% 
  pivot_longer(1:3) %>% 
  ggplot(aes(x = group, y = value, col = group)) +
  geom_point() +
  facet_wrap(.~name, scales = 'free_y') +
  scale_color_brewer(palette = "Set2") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

p1 <- anno.col2 %>% select(capture.rate, group) %>%
  ggplot(aes(x = group, y = capture.rate, col = group)) +
  # geom_boxplot() +
  geom_point(# position = position_jitter(width = 0.05, height = 0),
             size = 4) +
  scale_color_brewer(palette = "Set2") +
    labs(
    title = "Capture Rate in Chr12",
    x = "Group",
    y = "Proportion",
    color = "Group" # Clear legend title
  ) + 
  scale_y_continuous(limits = c(0, NA)) +
  geom_hline(yintercept = 1, linetype = 'dashed', col = 'grey70') +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        legend.position = "bottom")

p2 <- anno.col2 %>% select(count_DNM1L, group) %>%
  ggplot(aes(x = group, y = count_DNM1L/1e3, col = group)) +
  geom_point(# position = position_jitter(width = 0.2, height = 0),
             size = 4) +
  scale_color_brewer(palette = "Set2") +
    labs(
    title = "DNM1L Counts",
    x = "Group",
    y = "Counts (*1000)",
    color = "Group" # Clear legend title
  ) + 
  scale_y_continuous(limits = c(0, NA)) +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        legend.position = "bottom")

p3 <- anno.col2 %>% select(count_chr12, group) %>%
  ggplot(aes(x = group, y = count_chr12/1e3, col = group)) +
  geom_point(# position = position_jitter(width = 0.2, height = 0),
             size = 4) +
  scale_color_brewer(palette = "Set2") +
    labs(
    title = "Chr12 Total Counts",
    x = "Group",
    y = "Counts (*1000)",
    color = "Group" # Clear legend title
  ) +
  scale_y_continuous(limits = c(0, NA)) +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        legend.position = "bottom")

p <- plot_grid(p1,p2,p3, nrow = 1)
print(p)
ggsave('eda_plot/qc.pdf', width = 12, height = 5)
```

### PCA plots using log2CPM

```{r, dev=c('png', 'pdf')}
#| fig-cap: "Fig2A PCA plot using log2CPM of all transcripts on chr12"
#| layout-ncol: 2

exp2 <- log2(assays(se2)$CPM + 1) 
exp2 <- exp2[rowSums(exp2)!=0, ] # filter for non expressed

colnames(exp2) <- paste(group2, c('S1','S2'), sep = '_') # rename by replicate (S) 

autoplot(prcomp(exp2 %>% t()), 
         data = anno.col2,
         color = 'group', size = 'capture.rate') +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 

autoplot(prcomp(exp2 %>% t()), 
         data = anno.col2,
         color = 'group', size = 'count_chr12') +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 

autoplot(prcomp(exp2 %>% t()), 
         data = anno.col2,
         color = 'group', size = 'count_DNM1L') +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 

autoplot(prcomp(exp2 %>% t()), 
         data = anno.col2,
         color = 'group', size = 4) +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 
```


```{r, dev=c('png', 'pdf')}
#| fig-cap: "Fig2B PCA plot using only log2CPM of DNM1L"
#| layout-ncol: 2

exp2 <- log2(assays(se2)$CPM[rowData(se2)$GENEID == 'DNM1L',] + 1) 
exp2 <- exp2[rowSums(exp2)!=0, ] # filter for non expressed

# colnames(exp2) <- paste(group2, c('S1','S2'), sep = '_') # rename by replicate (S) 
colnames(exp2) <- c('CERA_A', 'CERA_B','CERA_C','LV_3160','LV5138','LV6086','LV3149', 'CM1','CM2','CM4')

autoplot(prcomp(exp2 %>% t()), 
         data = anno.col2,
         color = 'group', size = 'capture.rate') +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 

autoplot(prcomp(exp2 %>% t()), 
         data = anno.col2,
         color = 'group', size = 'count_chr12') +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 

autoplot(prcomp(exp2 %>% t()), 
         data = anno.col2,
         color = 'group', size = 'count_DNM1L') +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 

autoplot(prcomp(exp2 %>% t()), 
         data = anno.col2,
         color = 'group', size = 4) +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 
```
### Heatmap using log2CPM


```{r, fig.height=10, dev=c('png', 'pdf')}
#| fig-cap: "Fig3 Bambu output showing the structure and expression of isoforms"
plotBambu(se2, type = "annotation", gene_id = "DNM1L")

```

```{r, fig.height=7, dev=c('png', 'pdf')}
#| fig-cap: "Fig4 Heatmap showing isoform expression, clustered by default"

# anno.col2 <- data.frame(capture.rate = capture.rate2,
#                        group = group2,
#                        library.size = colSums(assays(se1)$counts))

rownames(anno.col2) <- colnames(exp2)

ann_colors <- list(group = setNames(brewer.pal(3, "Set2"), c("CERA","CL2","LV")),
                   capture.rate = brewer.pal(n = 9, name = "BuPu"),
                   count_DNM1L = brewer.pal(n = 9, name = "BuGn"))

exp2 %>%
  pheatmap::pheatmap(cellwidth = 15, cellheight = 15, 
           annotation_col = anno.col2[, 4, drop = F] ,
           annotation_colors = ann_colors,
           scale = 'none',
           show_colnames = F,
           main = 'Heatmap using log2(CPM+1)',
           width = 7, height = 6)

exp2 %>%
  pheatmap::pheatmap(cellwidth = 15, cellheight = 15, 
           annotation_col = anno.col2[, 4, drop = F] ,
           annotation_colors = ann_colors,
           scale = 'row',
           show_colnames = F,
           main = 'Heatmap using z-scaled log2(CPM+1)',
           width = 7, height = 6)
```

### Boxplot using log2CPM

```{r, fig.height=6, fig.width=14, dev=c('png', 'pdf')}
# 1) Align groups to exp2 columns (samples)
stopifnot(all(colnames(exp2) %in% rownames(anno.col2)))

# 2) Long-format: one row per (transcript, sample)
df_long <- as.data.frame(exp2) %>%
  rownames_to_column("transcript") %>%
  pivot_longer(-transcript, names_to = "sample", values_to = "log2CPM") %>%
  left_join(anno.col2 %>% rownames_to_column() %>% select(rowname, group), 
            by = c('sample' = 'rowname')) %>%
  mutate(transcript = factor(transcript, 
                             levels = c("NM_012062.5","NM_012063.4","NM_005690.5",
                                        "NM_001278463.2","NM_001278464.2",'NM_001278465.2',
                                        "NM_001278466.2", "NM_001330380.2", "XM_011520543.4",
                                        "XM_047428047.1")))

# 3) Boxplots of expression by group, faceted by transcript (one facet per row of exp2)
ggplot(df_long, aes(x = group, y = log2CPM, fill = group)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.3, size = 0.8) +
  scale_fill_brewer(palette = "Set2") +
  labs(x = NULL, y = "Transcript expression log2(CPM+1)") +
  theme_classic() +
  theme(legend.position = "none",
        strip.text = element_text(size = 9),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~ transcript, scales = "free_y", ncol = 5)  +
    force_panelsizes(rows = unit(2, 'in'),
                   cols = unit(2, 'in')) 

ggplot(df_long, aes(x = transcript, y = log2CPM, fill = group)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.3, size = 0.8) +
  scale_fill_brewer(palette = "Set2") +
  labs(x = NULL, y = "Transcript expression log2(CPM+1)") +
  theme_classic() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        strip.text = element_text(size = 9)) +
  facet_wrap(~ group, scales = "free_y")  +
    force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 
```

### PCA using log2TPM
This is only done because we need to caluclate TSI based on TPM, TPM can eliminate transcript length bias. Though in this data, the transcript length are quite similar, effects are minimal

```{r, fig.height=7, fig.width=14, dev=c('png', 'pdf')}

# Map samples to tissues (assumes colnames(exp2) are sample IDs)
stopifnot(all(colnames(exp2) %in% rownames(anno.col2)))

## get TPM first

calculateTPM <- function(se = se2) {
  
  counts <- assays(se)$counts %>% colSums()
  incompcounts <- data.frame(metadata(se)$incompatibleCounts)[, colnames(se)] %>% colSums()
  totalcounts <- counts + incompcounts
  
  tx_lengths <- rowRanges(se) %>% width() %>% sum
  
  stopifnot(rownames(counts) == names(tx_lengths))
  
  tpm <- assays(se)$counts/tx_lengths*1e3
  sum <- colSums(tpm)
  
  tmp <- tpm/sum*1e6
  return(tpm)

}

tpm <- calculateTPM(se2)
colnames(tpm) <- colnames(exp2)

# PCA 
autoplot(prcomp(log2(tpm+1) %>% t()), 
         data = anno.col2,
         color = 'group', size = 4) +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) +
  ggtitle('PCA using log2(TPM+1) for chr12 transcripts')

tpm <- tpm[rownames(exp2), ]

autoplot(prcomp(log2(tpm+1) %>% t()), 
         data = anno.col2,
         color = 'group', size = 4) +
  scale_color_brewer(palette = "Set2") +
  force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) +
  ggtitle('PCA using log2(TPM+1) of DNM1L transcripts')
```


### Heatmap using log2TPM
```{r, fig.height=7, fig.width=14, dev=c('png', 'pdf')}
# heatmap
log2(tpm + 1) %>%
  pheatmap::pheatmap(cellwidth = 15, cellheight = 15, 
           annotation_col = anno.col2[, 4, drop = F] ,
           annotation_colors = ann_colors,
           scale = 'none',
           show_colnames = F,
           main = 'Heatmap using log2(TPM+1)',
           width = 7, height = 6)

log2(tpm + 1) %>%
  pheatmap::pheatmap(cellwidth = 15, cellheight = 15, 
           annotation_col = anno.col2[, 4, drop = F] ,
           annotation_colors = ann_colors,
           scale = 'row',
           show_colnames = F,
           main = 'Heatmap using z-scaled log2(TPM+1)',
           width = 7, height = 6)
```


### Boxplot using log2TPM
```{r, fig.height=7, fig.width=14, dev=c('png', 'pdf')}
df_long2 <- as.data.frame(tpm) %>%
  rownames_to_column("transcript") %>%
  pivot_longer(-transcript, names_to = "sample", values_to = "TPM") %>%
  left_join(anno.col2 %>% rownames_to_column() %>% select(rowname, group), 
            by = c('sample' = 'rowname')) %>%
  mutate(transcript = factor(transcript, 
                             levels = c("NM_012062.5","NM_012063.4","NM_005690.5",
                                        "NM_001278463.2","NM_001278464.2",'NM_001278465.2',
                                        "NM_001278466.2", "NM_001330380.2", "XM_011520543.4",
                                        "XM_047428047.1")))

# 3) Boxplots of expression by group, faceted by transcript (one facet per row of exp2)
ggplot(df_long2, aes(x = group, y = log2(TPM+1), fill = group)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.3, size = 0.8) +
  scale_fill_brewer(palette = "Set2") +
  labs(x = NULL, y = "Transcript expression log2(TPM+1)") +
  theme_classic() +
  theme(legend.position = "none",
        strip.text = element_text(size = 9),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~ transcript, scales = "free_y", ncol = 5)  +
    force_panelsizes(rows = unit(2, 'in'),
                   cols = unit(2, 'in')) 

ggplot(df_long2, aes(x = transcript, y = log2(TPM+1), fill = group)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.3, size = 0.8) +
  scale_fill_brewer(palette = "Set2") +
  labs(x = NULL, y = "Transcript expression log2(TPM+1)") +
  theme_classic() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        strip.text = element_text(size = 9)) +
  facet_wrap(~ group, scales = "free_y")  +
    force_panelsizes(rows = unit(3, 'in'),
                   cols = unit(3, 'in')) 

# Mean TPM per (transcript, tissue), fill missing tissues with 0 to keep range [1/n, 1]
xi_tbl <- df_long2 %>%
  group_by(transcript, group) %>%
  summarise(xi = mean(TPM, na.rm = TRUE))

# TSI = max(xi) / sum(xi)
tsi_tbl <- xi_tbl %>%
  group_by(transcript) %>%
  summarise(
    sum_x = sum(xi),
    max_x = max(xi),
    TSI   = ifelse(sum_x > 0, max_x / sum_x, NA_real_),
    .groups = "drop"
  ) %>%
  arrange(desc(TSI))

print(tsi_tbl)


```


### Statistical test using log2CPM and log2TPM

Average expression as log2(CPM+1), TPM of each cell type, and ANOVA p values (both logCPM and logTPM).  

```{r, dev=c('png', 'pdf')}

pvals <- sapply(1:nrow(exp2), function(x){
  
  anova(lm(exp2[x,]~group2))$`Pr(>F)`[1]
  
}, simplify = T)

qvals <- p.adjust(pvals, method = 'BH')

pvals2 <- sapply(1:nrow(tpm), function(x){
  
  anova(lm(log2(tpm+1)[x,]~group2))$`Pr(>F)`[1]
  
}, simplify = T)

qvals2 <- p.adjust(pvals2, method = 'BH')

# Initialize an empty matrix to store the results
result2 <- matrix(NA, nrow = nrow(exp2), ncol = length(levels(group2)))
colnames(result2) <- levels(group2)
rownames(result2) <- rownames(exp2)

# Calculate the mean value in each row for each group
for (i in seq_along(levels(group2))) {
  group_samples <- which(group2 == levels(group2)[i])
  
  if (length(group_samples) == 1) {
    # If only one sample, use the value itself
    result2[, i] <- exp2[, group_samples]
  } else {
    # If multiple samples, calculate mean
    result2[, i] <- rowMeans(exp2[, group_samples], na.rm = TRUE)
  }
}

# Convert the result to a data frame for better readability
result2 <- as.data.frame(result2)

result2 <- data.frame(result2, pvals_lcpm = pvals, FDR_lcpm = qvals, 
                      pvals_ltpm = pvals2, FDR_ltpm = qvals2, 
                      exp2, 
                      tpm,
                      check.names = FALSE)

colnames(result2)[8:27] <- paste(colnames(result2)[8:27], 
                                 rep(c('log2CPM', 'TPM'), each = 10), 
                                 sep = '_') 

result2 <- left_join(result2 %>% rownames_to_column(var = 'transcript'),
                    tsi_tbl, by = 'transcript')

datatable(result2[c(1:8,29:31,9:28)]) %>%
  formatRound(columns=2:ncol(result2), digits=3)

write.csv(result2, 'human_anova.csv')

```

### Statistical analysis using edgeR QL test
```{r}
library(edgeR)
dge <- DGEList(counts = assays(se2)$counts, 
               genes = rowData(se2) %>% data.frame(),
               group = group2, 
               remove.zeros = T)

y <- normLibSizes(dge)
y$samples

## edgeR manual section 3.2.6
## ANOVA like test
design <- model.matrix(~group2, data = y$samples)
colnames(design)[-1] <- levels(group2)[-1]

y <- estimateDisp(y, design, robust = T)
y$common.dispersion
plotBCV(y)

fit <- glmQLFit(y, design, robust = T)
qlf <- glmQLFTest(fit, coef = 2:ncol(design))

deg <- topTags(qlf, n = Inf, p.value = Inf) %>% 
  data.frame() %>%
  filter(GENEID == 'DNM1L') 

# fold change here is relative to CERA
deg %>% 
  select(c(1:2, 12:ncol(deg))) %>% 
  datatable() %>%
  formatRound(columns=3:8, digits=3)

write.csv(deg[, -11], 'human_edger_anovalike.csv') # remove eqClassById list column
```



```{r}

sessionInfo()
```
